File: source/python/reference_counting.py
# Manual reference counting simulation

class RCObject:
    def __init__(self, data):
        self.refcount = 1
        self.data = data

def rc_new(data):
    obj = RCObject(data)
    print(f"created obj(data={data}, refcount=1)")
    return obj

def rc_retain(obj):
    obj.refcount += 1
    print(f"retain obj(data={obj.data}, refcount={obj.refcount})")
    return obj

def rc_release(obj):
    obj.refcount -= 1
    if obj.refcount == 0:
        print(f"release obj(data={obj.data}, refcount=0) -> freed")
        return None
    print(f"release obj(data={obj.data}, refcount={obj.refcount})")
    return obj

print("--- create object ---")
a = rc_new(42)

print("--- share with b ---")
b = rc_retain(a)

print("--- share with c ---")
c = rc_retain(a)

print(f"data: {a.data}, refcount: {a.refcount}")

print("--- release a ---")
a = rc_release(a)

print("--- release b ---")
b = rc_release(b)

print("--- release c (last ref, should free) ---")
c = rc_release(c)

File: source/javascript/reference_counting.js
// Manual reference counting simulation
function rcNew(data) {
    const obj = { refcount: 1, data };
    console.log(`created obj(data=${data}, refcount=1)`);
    return obj;
}

function rcRetain(obj) {
    obj.refcount++;
    console.log(`retain obj(data=${obj.data}, refcount=${obj.refcount})`);
    return obj;
}

function rcRelease(obj) {
    obj.refcount--;
    if (obj.refcount === 0) {
        console.log(`release obj(data=${obj.data}, refcount=0) -> freed`);
        return null;
    }
    console.log(`release obj(data=${obj.data}, refcount=${obj.refcount})`);
    return obj;
}

console.log("--- create object ---");
let a = rcNew(42);
console.log("--- share with b ---");
let b = rcRetain(a);
console.log("--- share with c ---");
let c = rcRetain(a);
console.log(`data: ${a.data}, refcount: ${a.refcount}`);
console.log("--- release a ---");
a = rcRelease(a);
console.log("--- release b ---");
b = rcRelease(b);
console.log("--- release c (last ref, should free) ---");
c = rcRelease(c);
